stl/inc/xstring

```diff
@@ -2642,6 +2642,47 @@ private:
     }
 
 public:
+#if _HAS_CXX20
+    void _Move_assign_from_buffer(_Elem* const _Right, const size_type _Size, const size_type _Res) {
+        // Move assign from a buffer, used by basic_stringbuf
+        _Tidy_deallocate();
+        _Traits::assign(_Right[_Size], _Elem());
+        pointer _Fancy_right = pointer_traits<pointer>::pointer_to(*_Right);
+        auto& _My_data       = _Mypair._Myval2;
+        _My_data._Mysize     = _Size;
+        _My_data._Myres      = _Res;
+        if (_My_data._Large_string_engaged()) {
+            _Construct_in_place(_My_data._Bx._Ptr, _Fancy_right);
+        } else {
+            _Traits::copy(_My_data._Bx._Buf, _Right, _Res);
+            _Getal().deallocate(_Fancy_right, _Res);
+        }
+    }
+
+    struct _Released_buffer {
+        pointer _Ptr;
+        size_type _Size;
+        size_type _Res;
+        allocator_type& _Al;
+    };
+
+    _NODISCARD _Released_buffer _Release_to_buffer() {
+        // Release to a buffer, or allocate a new one if in small string mode
+        auto& _My_data = _Mypair._Myval2;
+        pointer _Result_ptr;
+        size_type _Result_size = _My_data._Mysize;
+        size_type _Result_res  = _My_data._Myres;
+        if (_My_data._Large_string_engaged()) {
+            _Construct_in_place(_Result_ptr, _STD move(_My_data._Bx._Ptr));
+            _Tidy_init();
+        } else {
+            _Construct_in_place(_Result_ptr, _STD move(_Getal().allocate(_BUF_SIZE)));

```

Comment:
Maybe this should use `_BUF_SIZE` + 1 so that moving to and from an `sstream` doesn't repeatedly trigger allocation and deallocation of the buffer? That suggests a test like the following pseudocode:

```
string s;
s.push_back('a');
sstream example(std::move(s));
s = std::move(example).str();
example.str(std::move(s));
assert(number of allocations == 1);
```

Reply:
I'll try it, but how to get the number of allocations?

Reply:
>I'll try it, but how to get the number of allocations?

Write an allocator that counts.